aspnet/mvc/controllers/testing.rst

```diff
@@ -1,8 +1,158 @@
 
-.. include:: /../common/stub-notice.txt
+.. contents:: Sections
+	:local:
+	:depth: 1
+	
+`View sample files <https://github.com/aspnet/Docs/tree/1.0.0-rc1/aspnet/mvc/controllers/testing/sample>`_
 
-.. _issue: https://github.com/aspnet/Docs/issues/123
+What is Controller Logic
+------------------------
+*Controllers* define groups of *actions*. The framework could have been designed around isolated actions, but grouping them using controllers often useful for :doc:`routing </fundamentals/routing>`, applying :doc:`filters <filters>`, and :doc:`injecting common dependencies <dependency-injection>`. 
+
+:doc:`Learn more about controllers and actions <actions>`.
+
+Controller logic should be minimal and should not be focused on business logic or infrastructure concerns like data access. When testing controller logic, avoid testing the framework as much as possible. You should test how the controller *behaves* based on valid or invalid inputs, and what kind of response it provides based on the result of some business operation it performs.
+
+Typical controller responsibilities:
+	- Verify ``ModelState.IsValid``
+	- Return an error response if ``ModelState`` is invalid
+	- Retrieve a business entity from persistence
+	- Perform an action on the business entity
+	- Save the business entity to persistence
+	- Return an appropriate ``IActionResult``
+
+Unit Testing
+------------
+:doc:`Unit testing </testing/unit-testing>` involves testing a part of an app in isolation from its infrastructure and dependencies. When unit testing controller logic, only the contents of a single action should be tested, not the behavior of its dependencies or of the framework itself. Thus, when unit testing a controller's actions, avoid testing global or attribute-based :doc:`filters <filters>`, :doc:`routing </fundamentals/routing>`, or :doc:`model binding </mvc/models/model-binding>`, as these are all performed by the framework (testing these is an integration test responsibility).
+
+To demonstrate unit testing, review the following controller. It displays a list of brainstorming sessions and allows new brainstorming sessions to be created with a POST:
+
+.. literalinclude:: testing/sample/TestingControllersSample/src/TestingControllersSample/Controllers/HomeController.cs
+  :language: c#
+  :emphasize-lines: 11,15,20,40-41
+
+The controller is following the `explicit dependencies principle <http://deviq.com/explicit-dependencies-principle/>`_, expecting dependency injection to provide it with an instance of ``IBrainStormSessionRepository``. This makes it fairly easy to test using a mock object framework, like `Moq <https://www.nuget.org/packages/Moq/>`_. To test the first ``Index`` method, which has no looping or branching logic and only calls one method, we need only verify that a ``ViewResult`` is returned, with a ``ViewModel`` containing whatever was returned from the repository's ``List`` method.
```

Comment:
The ``HTTP GET Index`` method has no looping or branching logic and only calls one method. To test this ``Index`` method,  we need to verify that a ``ViewResult`` is returned, with a ``ViewModel`` from the repository's ``List`` method.